bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/composite/AbstractCompositeModuleHandler.java

```diff
@@ -88,10 +89,14 @@ public AbstractCompositeModuleHandler(M module, MT moduleType, LinkedHashMap<M,
 
     @Override
     public void dispose() {
-        if (moduleHandlerMap != null) {
-            moduleHandlerMap.clear();
-            moduleHandlerMap = null;
+        List<M> children = getChildren();
+        for (M child : children) {
+            ModuleHandler childHandler = moduleHandlerMap.remove(child);
+            childHandler.dispose();
         }
```

Comment:
I'd be ok with your option 2 (BaseTriggerHandler sets callback to null in dispose) as this at least shows developers that this is needed. But it still means that your PR needs to fix GenericEventTriggerHandler as it does not even call super...